src/pages/api/proxy/[...path].ts

```diff
@@ -3,23 +3,90 @@ import { EventEmitter } from 'events';
 import { createProxyMiddleware, fixRequestBody } from 'http-proxy-middleware';
 import { NextApiRequest, NextApiResponse } from 'next';
 
-import { X_AUTH_SIGNATURE, X_INTERNAL_CLIENT, X_TIMESTAMP } from '@/api';
 import generateSignature from '@/utils/auth/signature';
+import {
+  X_AUTH_SIGNATURE,
+  X_INTERNAL_CLIENT,
+  X_TIMESTAMP,
+  X_PROXY_SIGNATURE,
+  X_PROXY_TIMESTAMP,
+} from '@/utils/headers';
 
-// Define error messages in a constant object
 const ERROR_MESSAGES = {
   PROXY_ERROR: 'Proxy error',
   PROXY_HANDLER_ERROR: 'Proxy handler error',
+  FORBIDDEN: 'Forbidden',
 };
 
+const ALLOWED_DOMAINS = (process.env.ALLOWED_ORIGINS || '')
+  .split(',')
+  .map((domain) => domain.trim());
+
 // This line increases the default maximum number of event listeners for the EventEmitter to a better number like 20.
 // It is necessary to prevent memory leak warnings when multiple listeners are added,
 // which can occur in a proxy setup like this where multiple requests are handled concurrently.
 EventEmitter.defaultMaxListeners = Number(process.env.PROXY_DEFAULT_MAX_LISTENERS) || 100;
 
-// This file sets up a proxy middleware for API requests. It is needed to forward requests from the frontend
-// to the backend server, allowing for features like cookie handling and request body fixing, which are essential
-// for maintaining session state and ensuring correct request formatting while in a cross domain env.
+const isOriginAllowed = (origin: string | undefined): boolean => {
+  if (!origin) return false;
+  const url = new URL(origin);
+  const { hostname } = url;
+  return ALLOWED_DOMAINS.some((domain) => hostname.endsWith(domain));
```

Comment:
Why do we need `endsWith` here? `ALLOWED_DOMAINS` will be exact match + all subdomains already, I believe we should do `return ALLOWED_DOMAINS.includes(hostname);` issue with `endsWith` is that any it would accept any domain that happens to end with one of the configured ones for example `copy-quran.com` which we don't want.